### PR TITLE
Block helper functions for successful function calls

### PR DESCRIPTION
**Pull request by gabehamilton**
See Feature Request https://github.com/near/queryapi/issues/366
 * successfulFunctionCalls
 * socialOperations
    This is more specific than other methods in the class. However, this specificity helps users. In future we could add additional "more specific" helper methods for fungible tokens and NFTs.

Should we add Wrapper types for all Action Operations? see https://github.com/near/queryapi/issues/409

### CODE REVIEW

File: packages/near-lake-primitives/src/types/block.ts [hunk 1]
```diff
@@ -157,6 +166,90 @@ export class Block {
...
+                !functionCall.args.data ||
+                !Object.keys(functionCall.args.data) ||
+                !Object.keys(functionCall.args.data)[0]
+              ) {
+                  console.error(
+                    "Set operation did not have arg data in expected format"
+                  );
+                  return;
+              }
+              const accountId = Object.keys(functionCall.args.data)[0];
```
Comment (morgsmccauley) [inline packages/near-lake-primitives/src/types/block.ts:230]:
```suggestion
              const accountId = Object.keys(functionCall.args.data ?? {})[0];
              if (!accountId) {
                  return false;
              }
```

I believe this is the same?

Same comment around logging, can we please remove? Also we should be returning `false` here and below?

File: packages/near-lake-primitives/src/types/block.ts [hunk 2]
```diff
@@ -157,6 +166,90 @@ export class Block {
...
+          );
+    };
+
+    /**
+     *  Returns data that follows the social.near contract template for key value data.
+     * @param operation The top level key to search for in each user's account space
+     * @param contract Defaults to 'social.near', pass in a different contract if needed
+     * @param method Defaults to 'set', pass in a different method if needed
+     */
+    socialOperations(operation: string, contract: string = "social.near", method: string = "set") {
```
Comment (morgsmccauley) [inline packages/near-lake-primitives/src/types/block.ts:215]: Can we add a return type here please?

File: packages/near-lake-primitives/test/block.test.ts [hunk 3]
```diff
@@ -75,4 +75,24 @@ describe("Block", () => {
     );
     expect(authorToPostId).toMatchSnapshot();
   });
+
+  it("parses successful function calls", async () => {
```
Comment (morgsmccauley) [inline packages/near-lake-primitives/test/block.test.ts:79]: Can we add a test for `actions(successfulOnly)` too please?

File: packages/near-lake-primitives/src/types/receipts.ts [hunk 4]
```diff
@@ -167,7 +169,7 @@ class DeleteAccount {
 export type Operation =
   | 'CreateAccount'
   | DeployContract
-  | FunctionCall
```
Comment (morgsmccauley) [inline packages/near-lake-primitives/src/types/receipts.ts:170]: Was this incorrect? Is this a breaking change?

**Review by morgsmccauley — CHANGES_REQUESTED** [review]

File: packages/near-lake-primitives/src/types/block.ts [hunk 5]
```diff
@@ -79,11 +85,14 @@ export class Block {
...
      */
-    actions(): Action[] {
+    actions(successfulOnly?: boolean): Action[] {
         const actions: Action[] = this.streamerMessage.shards
             .flatMap((shard) => shard.receiptExecutionOutcomes)
-            .filter((exeuctionOutcomeWithReceipt) => Action.isActionReceipt(exeuctionOutcomeWithReceipt.receipt))
-            .map((exeuctionOutcomeWithReceipt) => Action.fromReceiptView(exeuctionOutcomeWithReceipt.receipt))
+            .filter((executionOutcomeWithReceipt) =>
+                Action.isActionReceipt(executionOutcomeWithReceipt.receipt) &&
+                (!successfulOnly || (executionOutcomeWithReceipt.executionOutcome.outcome.status as {SuccessValue: any}))
```
[the remark this answers is not in the document]
Comment (gabehamilton) [reply packages/near-lake-primitives/src/types/block.ts:93]: @pavelisnear hopefully your function is better than mine here.

File: packages/near-lake-primitives/src/types/receipts.ts (hunk 4 above)
Comment (gabehamilton) [reply packages/near-lake-primitives/src/types/receipts.ts:170]: @pavelisnear This file is where I ran into the Operations types not matching the data. I only changed FunctionCall but they should probably all be corrected.

**Comment by morgsmccauley** [comment]
Should we close now that https://github.com/near/near-lake-framework-js/pull/61 is merged?

**Comment by gabehamilton** [comment]
Yes, closing, #61 looks great. I do think more helper functions in either near-lake or queryapi would still be useful, extracting: SocialDB operations, NFTs, FTs, DAO proposals. Hiding all the error handling so that indexers can be just a few lines will make for great demos and onboarding.

**Pull request closed without merging**